train/trainer.py

```diff
@@ -254,27 +260,58 @@ def reconfigure_optimizers(self):
         optimizers, _ = self.configure_optimizers(
             existing_scheduler=self.lr_schedulers())
         self.optimizers()._optimizer = optimizers[0]
-
+    
+    def get_optimizer_params(self, weight_decay):
+        decay_params = []
+        no_decay_params = []
+        for name, param in self.model.named_parameters():
+            if not param.requires_grad:
+                continue  # Exclude frozen parameters
+            if "bias" in name or "LayerNorm.weight" in name or "LayerNorm.bias" in name:
+                no_decay_params.append(param)
+            else:
+                decay_params.append(param)
+        return [
+            {'params': decay_params, 'weight_decay': weight_decay},
+            {'params': no_decay_params, 'weight_decay': 0.0}
+        ]
+    
     def configure_optimizers(self, existing_scheduler=None):
-        optimizer = torch.optim.Adam(self.parameters(),
-                                     lr=self.train_params.lr)
+        weight_decay = self.train_params.weight_decay
 
-        def f_warmup(n):
-            assert n <= self.train_params.lr_warm_steps
-            return n / self.train_params.lr_warm_steps
+        if weight_decay > 0:
+            optimizer_grouped_parameters = self.get_optimizer_params(weight_decay)
+            optimizer = AdamW(optimizer_grouped_parameters, lr=self.train_params.lr)
+        else:
+            optimizer = torch.optim.Adam(self.parameters(), lr=self.train_params.lr)
 
+        def f_warmup(n):
```

Comment:
stylistic preference, but please undo the change to f_warmup (unless you have a particular reason for it?). i like code to fail loudly if something unexpected happens (e.g. if i call f_warmup after the expected warmup period)